File: cognipy/CogniPyCLI/InteractiveMode.cs
using CogniPy;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Text;

namespace CogniPyCLI
{
    class InteractiveMode
    {
         public static void EntryPoint(string[] args)
        {
            Stream inputStream = Console.OpenStandardInput();
            Stream outputStream = Console.OpenStandardOutput();

            var reader = new StreamReader(inputStream);
            var writer = new StreamWriter(outputStream);

            Dictionary<string, CogniPySvr> clients = new Dictionary<string, CogniPySvr>();

            while (true)
              {
                var cmd = reader.ReadLine();

                if (cmd == "@create")
                {
                    var uid = Guid.NewGuid().ToString();
                    var fe = new CogniPySvr();
                    clients.Add(uid, fe);
                    writer.Write(uid);
                }
                else if (cmd == "@delete")
                {
                    var uid = reader.ReadLine();
                    clients.Remove(uid);
                    writer.Write("@deleted");
                }
                else if (cmd == null || cmd == "@exit")
                {
                    break;
                }
                else
                {
                    try
                    {
                        var uid = reader.ReadLine();
                        var fe = clients[uid];
                        StringBuilder sb = new StringBuilder();
                        while (true)
                        {
                            var line = reader.ReadLine();
                            if (line == "\0")
                                break;
                            sb.Append(line);
                        }
                        JsonSerializer serializer = JsonSerializer.Create(new JsonSerializerSettings
                        {
                            Formatting = Newtonsoft.Json.Formatting.Indented,
                            ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore
                        });
                        var parms = serializer.Deserialize<object[]>(new JsonTextReader(new StringReader(sb.ToString())));
                        object ret = null;
                        try
                        {
                            var method = fe.GetType().GetMethod(cmd);
                            var ptp = method.GetParameters();
                            var cps = new List<object>();
                            for (var idx = 0; idx < ptp.Length; idx++)
                            {
                                object cp;
                                if (parms[idx] is JToken)
                                    cp = (parms[idx] as JToken).ToObject(ptp[idx].ParameterType);
                                else
                                    cp = parms[idx];
                                cps.Add(cp);
                            }

                            ret = method.Invoke(fe, cps.ToArray());
                        }
                        catch (AmbiguousMatchException)
                        { 
                            ret = fe.GetType().InvokeMember(cmd, BindingFlags.DeclaredOnly |
                                                               BindingFlags.Public | BindingFlags.NonPublic |
                                                               BindingFlags.Instance | BindingFlags.InvokeMethod,
                                null, fe, parms);
                        }

                        writer.WriteLine("@result");
                        serializer.Serialize(new JsonTextWriter(writer), ret);
                    }
                    catch (Exception ex)
                    {
                        if (ex is TargetInvocationException)
                            ex = ex.InnerException;

                        writer.WriteLine("@exception");
                        var w = new JsonTextWriter(writer);
                        JsonSerializer serializer = JsonSerializer.Create(new JsonSerializerSettings
                        {
                            Formatting = Newtonsoft.Json.Formatting.Indented,
                            ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore
                        });
                        serializer.Serialize(w,new object[] { ex.GetType().Name, ex });
                    }
                }
                writer.WriteLine();
                writer.WriteLine("\0");
                writer.Flush();
            }
        }
    }
}

File: cognipy/cognipycli/cognipycli/InteractiveMode.java
package cognipycli;

import CogniPy.*;
import Newtonsoft.Json.*;
import Newtonsoft.Json.Linq.*;
import java.util.*;
import java.io.*;

public class InteractiveMode
{
	public static void EntryPoint(String[] args)
	{
		InputStream inputStream = Console.OpenStandardInput();
		OutputStream outputStream = Console.OpenStandardOutput();

		InputStreamReader reader = new InputStreamReader(inputStream);
		OutputStreamWriter writer = new OutputStreamWriter(outputStream);

		HashMap<String, CogniPySvr> clients = new HashMap<String, CogniPySvr>();

		while (true)
		{
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java unless the Java 10 inferred typing option is selected:
			var cmd = reader.ReadLine();

			if (cmd.equals("@create"))
			{
				String uid = UUID.NewGuid().toString();
				CogniPySvr fe = new CogniPySvr();
				clients.put(uid, fe);
				writer.write(uid);
			}
			else if (cmd.equals("@delete"))
			{
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java unless the Java 10 inferred typing option is selected:
				var uid = reader.ReadLine();
				clients.remove(uid);
				writer.write("@deleted");
			}
			else if (cmd == null || cmd.equals("@exit"))
			{
				break;
			}
			else
			{
				try
				{
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java unless the Java 10 inferred typing option is selected:
					var uid = reader.ReadLine();
					CogniPySvr fe = clients.get(uid);
					StringBuilder sb = new StringBuilder();
					while (true)
					{
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java unless the Java 10 inferred typing option is selected:
						var line = reader.ReadLine();
						if (line.equals("\0"))
						{
							break;
						}
						sb.append(line);
					}
					JsonSerializer serializer = new JsonSerializer();
					Object parms = serializer.<Object[]>Deserialize(new JsonTextReader(new StringReader(sb.toString())));
					Object ret = null;
					try
					{
						java.lang.reflect.Method method = fe.getClass().getMethod(cmd);
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java unless the Java 10 inferred typing option is selected:
						var ptp = method.GetParameters();
						ArrayList<Object> cps = new ArrayList<Object>();
						for (int idx = 0; idx < ptp.Length; idx++)
						{
							Object cp;
							if (parms[idx] instanceof JToken)
							{
								cp = (parms[idx] instanceof JToken ? (JToken)parms[idx] : null).ToObject(ptp[idx].ParameterType);
							}
							else
							{
								cp = parms[idx];
							}
							cps.add(cp);
						}

						ret = method.Invoke(fe, cps.toArray(new Object[0]));
					}
					catch (AmbiguousMatchException e)
					{
						ret = fe.getClass().InvokeMember(cmd, BindingFlags.DeclaredOnly.getValue() | BindingFlags.Public.getValue() | BindingFlags.NonPublic.getValue() | BindingFlags.Instance.getValue() | BindingFlags.InvokeMethod.getValue(), null, fe, parms);
					}

					writer.write("@result" + System.lineSeparator());
					serializer.Serialize(new JsonTextWriter(writer), ret);
				}
				catch (RuntimeException ex)
				{
					writer.write("@exception" + System.lineSeparator());
					JsonTextWriter w = new JsonTextWriter(writer);
					JsonSerializer serializer = new JsonSerializer();
					serializer.Serialize(w, ex);
				}
			}
			writer.WriteLine();
			writer.write("\0" + System.lineSeparator());
			writer.flush();
		}
	}
}